File: src/rank_bv.rs
pub use cxx::UniquePtr;
pub use rbv::*;

#[cxx::bridge]
mod rbv {
    unsafe extern "C++" {
        include!("CBL/cxx/rank_bv.h");

        type RankBV;
        fn new_rank_bv(size: usize) -> UniquePtr<RankBV>;
        fn size(&self) -> usize;
        fn get(&self, index: usize) -> bool;
        fn set(&self, index: usize) -> bool;
        fn clear(&self, index: usize) -> bool;
        fn toggle(&self, index: usize) -> bool;
        fn rank(&self, index: usize) -> u64;
        fn count_ones(&self) -> usize;
    }
}

File: cxx/rank_bv.cc
#include "rank_bv.h"

RankBV::RankBV(size_t size) : bitvect((size + 63) / 64), bv(&bitvect, size) {}

RankBV::~RankBV() {}

size_t RankBV::size() const { return bv.size(); }

bool RankBV::get(size_t index) const {
  return bv.bitvector()[index / 64] & (1ULL << (index % 64));
}

bool RankBV::set(size_t index) const { return bv.set(index); }

bool RankBV::clear(size_t index) const { return bv.clear(index); }

bool RankBV::toggle(size_t index) const { return bv.toggle(index); }

uint64_t RankBV::rank(size_t index) const { return bv.rank(index); }

size_t RankBV::count_ones() const { return bv.rank(bv.size() - 1); }

unique_ptr<RankBV> new_rank_bv(size_t size) {
  return unique_ptr<RankBV>(new RankBV(size));
}
